### PR TITLE
[MM] Pad Qwen2_5_VisionMLP weights to use CUTLASS kernels

### PR DESCRIPTION
**Pull request by huanghua1994**
## Purpose

This PR optimizes Qwen2.5-VL FP8 prefill performance by padding the shape of multiple linear layers to multiplies of 16 for using fast CUTLASS kernels instead of falling back to Triton `scaled_mm_kernel`.

Previously, #41868 (partially) solved this issue by padding both input matrices in `CutlassFP8ScaledMMLinearKernel`. This approach has potential side efforts on other models.

This PR targets only the vision MLP layer in Qwen2.5-VL. It zero-pads the up & gate & down projection weight matrices and avoids de-padding gate & up projection outputs and padding for down projection input.

## Test Plan

Launch engine:

```bash
export MM_LIMIT='{"image": 1}'
export COMPILATION_CONFIG='{"compile_mm_encoder": true, "cudagraph_mode": "FULL_AND_PIECEWISE", "max_cudagraph_capture_size": 2048}'
python -m vllm.entrypoints.openai.api_server \
  --model Qwen/Qwen2.5-VL-7B-Instruct \
  --profiler-config.profiler cuda \
  --no-enable-prefix-caching \
  --max-model-len 32768 \
  --max-num-seqs 1024 \
  --distributed-executor-backend uni \
  --limit-mm-per-prompt "$MM_LIMIT" \
  --compilation-config "$COMPILATION_CONFIG" \
  --max-num-batched-tokens 2048 \
  --quantization fp8
```

Benchmark:

```bash
vllm bench serve \
  --backend openai-chat \
  --model Qwen/Qwen2.5-VL-7B-Instruct \
  --endpoint /v1/chat/completions \
  --base-url http://127.0.0.1:8000 \
  --profile \
  --metric-percentiles 50,90,99 \
  --dataset-name random-mm \
  --random-mm-base-items-per-request 1 \
  --random-mm-num-mm-items-range-ratio 0 \
  --random-mm-bucket-config "{(512, 512, 1): 1.0}" \
  --random-mm-limit-mm-per-prompt '{"image": 1}' \
  --input-len 512 \
  --output-len 100 \
  --num-warmups 50 \
  --num-prompts 1000 \
  --request-rate 100 
```

## Test Result

On a single B200 GPU, without this PR:

```text
============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Request rate configured (RPS):           100.00
Benchmark duration (s):                  14.68
Total input tokens:                      856999
Total generated tokens:                  100000
Request throughput (req/s):              68.13
Output token throughput (tok/s):         6813.09
Peak output token throughput (tok/s):    12305.00
Peak concurrent requests:                587.00
Total token throughput (tok/s):          65201.25
---------------Time to First Token----------------
Mean TTFT (ms):                          2151.64
Median TTFT (ms):                        2115.74
P50 TTFT (ms):                           2115.74
P90 TTFT (ms):                           3851.23
P99 TTFT (ms):                           4219.73
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          27.92
Median TPOT (ms):                        30.59
P50 TPOT (ms):                           30.59
P90 TPOT (ms):                           31.71
P99 TPOT (ms):                           31.84
---------------Inter-token Latency----------------
Mean ITL (ms):                           27.65
Median ITL (ms):                         29.58
P50 ITL (ms):                            29.58
P90 ITL (ms):                            35.84
P99 ITL (ms):                            45.15
==================================================
```

With this PR:

```text
============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Request rate configured (RPS):           100.00
Benchmark duration (s):                  14.58
Total input tokens:                      856999
Total generated tokens:                  100000
Request throughput (req/s):              68.57
Output token throughput (tok/s):         6856.56
Peak output token throughput (tok/s):    11739.00
Peak concurrent requests:                579.00
Total token throughput (tok/s):          65617.16
---------------Time to First Token----------------
Mean TTFT (ms):                          2060.89
Median TTFT (ms):                        2020.54
P50 TTFT (ms):                           2020.54
P90 TTFT (ms):                           3768.02
P99 TTFT (ms):                           4137.64
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          27.85
Median TPOT (ms):                        30.41
P50 TPOT (ms):                           30.41
P90 TPOT (ms):                           31.37
P99 TPOT (ms):                           31.48
---------------Inter-token Latency----------------
Mean ITL (ms):                           27.57
Median ITL (ms):                         29.63
P50 ITL (ms):                            29.63
P90 ITL (ms):                            35.46
P99 ITL (ms):                            42.74
==================================================
```

5% faster TTFT.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -35,6 +35,7 @@
 import torch
```
[the remark this answers is not in the document]
Comment (huanghua1994) [reply vllm/model_executor/models/qwen2_5_vl.py:35]: Fixed as suggested.

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 2]
```diff
@@ -119,6 +142,590 @@
...
+                and self.tp_size > 1
+            ):
+                raise NotImplementedError(
+                    "Shard id with multiple indices is not supported "
+                    "for BNB quantization with TP yet."
+                )
+            shard_offsets: list[tuple[int, int, int]] = []
+            for i, output_size in enumerate(output_sizes):
+                shard_offsets.append((i, current_shard_offset, output_size))
+                current_shard_offset += output_size
```
[the remark this answers is not in the document]
Comment (huanghua1994) [reply vllm/model_executor/models/qwen2_5_vl.py]: Fixed as suggested.

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 3]
```diff
@@ -119,6 +142,590 @@
...
+        param: BasevLLMParameter,
+        loaded_weight: torch.Tensor,
+        output_sizes: list[int] | None = None,
+    ):
+        current_shard_offset = 0
+        shard_offsets: list[tuple[int, int, int]] = []
+        output_sizes = output_sizes or self.loaded_output_sizes
+        for i, output_size in enumerate(output_sizes):
+            shard_offsets.append((i, current_shard_offset, output_size))
+            current_shard_offset += output_size
```
[the remark this answers is not in the document]
Comment (huanghua1994) [reply vllm/model_executor/models/qwen2_5_vl.py]: Fixed as suggested.

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 4]
```diff
@@ -119,6 +142,590 @@
...
+            )
+            if isinstance(param, BlockQuantScaleParameter):
+                weight_block_size = getattr(self, "weight_block_size", None)
+                output_sizes = [
+                    adjust_block_scale_shard(weight_block_size, size, 0)[0]
+                    for size in (output_sizes or self.loaded_output_sizes)
+                ]
+            self._load_fused_module_from_checkpoint(
+                param, loaded_weight, output_sizes=output_sizes
+            )
```
[the remark this answers is not in the document]
Comment (huanghua1994) [reply vllm/model_executor/models/qwen2_5_vl.py]: Fixed as suggested.

**Comment by johncalesp** [comment]
Hi @huanghua1994  in general I have some concerns about overwriting `weight_loader` and `weight_loader_v2` , because future modifications to these loaders will be missed (new quant types, params).

I was wondering if we can consider modifying the weights after loading and applying the quantization.
Also we need to keep into account that there are other quantization ( GPTQ, AWQ, Marlin) and I'm not sure if and how we are affecting their weights loading. So we may need to gate these changes to only apply them for the flag `--quantization fp8`

something like:
```
def _round_up_to_multiple(value: int, multiple: int) -> int:
    return ((value + multiple - 1) // multiple) * multiple

def _is_fp8(quant_config) -> bool:
    return quant_config is not None and getattr(quant_config, "quant_method", None) == "fp8"


def _pad_weight_per_shard(
    layer: nn.Module,
    *,
    per_shard_output_pad: int,
    input_pad: int,
) -> None:
    """Zero-pad ``layer.weight`` and ``layer.weight_scale`` in place.
    Output padding is applied PER OUTPUT SHARD (so merged gate/up are padded
    independently — concatenating naively would shift the up shard). Input
    padding is applied along the per-rank input dim.
    """
    if per_shard_output_pad == 0 and input_pad == 0:
        return
    w = layer.weight.data  # [out_per_partition, in_per_partition]
    num_shards = len(layer.output_partition_sizes)
    per_shard_out = layer.output_size_per_partition // num_shards
    shards = w.split(per_shard_out, dim=0)
    shards = [F.pad(s, (0, input_pad, 0, per_shard_output_pad)) for s in shards]
    layer.weight = Parameter(torch.cat(shards, dim=0), requires_grad=False)
    new_per_shard = per_shard_out + per_shard_output_pad
    layer.output_partition_sizes = [new_per_shard] * num_shards
    layer.output_size_per_partition = new_per_shard * num_shards
    layer.input_size_per_partition += input_pad
    # Pad per-channel weight_scale along the output dim with 1.0 (safe because
    # the corresponding weight rows are exactly zero).
    ws = getattr(layer, "weight_scale", None)
    if isinstance(ws, torch.nn.Parameter) and ws.dim() >= 1 and ws.shape[0] > 1:
        ws_shards = ws.data.split(per_shard_out, dim=0)
        ws_shards = [F.pad(s, (0, per_shard_output_pad), value=1.0) for s in ws_shards]
        layer.weight_scale = Parameter(torch.cat(ws_shards, dim=0), requires_grad=False)


def _install_post_pad_hook(
    layer: nn.Module,
    *,
    per_shard_output_pad: int,
    input_pad: int,
) -> None:
    """Wrap layer.quant_method.process_weights_after_loading so it pads after
    the quant method finishes (e.g. after FP8 quantization)."""
    qm = layer.quant_method
    original = qm.process_weights_after_loading
    def wrapped(module: nn.Module) -> None:
        original(module)
        _pad_weight_per_shard(
            module,
            per_shard_output_pad=per_shard_output_pad,
            input_pad=input_pad,
        )
    qm.process_weights_after_loading = wrapped  # type: ignore[method-assign]


class Qwen2_5_VisionMLP(nn.Module):
    def __init__(
        self,
        in_features: int,
        hidden_features: int,
        bias: bool = False,
        act_fn: Callable[[torch.Tensor], torch.Tensor] = F.silu,
        quant_config: QuantizationConfig | None = None,
        prefix: str = "",
    ):
        super().__init__()
        use_data_parallel = is_vit_use_data_parallel()
        tp_size = (
            1
            if use_data_parallel
            else get_tensor_model_parallel_world_size()
        )
        self.in_features = in_features
        self.hidden_features = hidden_features
        # Layers are constructed with the ORIGINAL sizes — base loader runs
        # unchanged, divide() still asserts hidden_features % tp_size == 0.
        self.gate_up_proj = MergedColumnParallelLinear(
            input_size=in_features,
            output_sizes=[hidden_features] * 2,
            bias=bias,
            quant_config=quant_config,
            prefix=f"{prefix}.gate_up_proj",
            disable_tp=use_data_parallel,
        )
        self.down_proj = RowParallelLinear(
            hidden_features,
            in_features,
            bias=bias,
            quant_config=quant_config,
            prefix=f"{prefix}.down_proj",
            disable_tp=use_data_parallel,
        )
        self.act_fn = act_fn
        # Pad targets: each per-rank dimension touched by a CUTLASS GEMM must
        # be a multiple of 16. Pad input/output independently per shard.
        self.padded_in_features = _round_up(in_features, 16)
        per_shard_hidden = hidden_features // tp_size
        padded_per_shard_hidden = _round_up(per_shard_hidden, 16)
        self.padded_hidden_per_shard = padded_per_shard_hidden
        if _is_fp8(quant_config):
            _install_post_pad_hook(
                self.gate_up_proj,
                per_shard_output_pad=padded_per_shard_hidden - per_shard_hidden,
                input_pad=self.padded_in_features - in_features,
            )
            _install_post_pad_hook(
                self.down_proj,
                per_shard_output_pad=self.padded_in_features - in_features,
                input_pad=padded_per_shard_hidden - per_shard_hidden,
            )
    
      def forward(self, x: torch.Tensor):
        if x.shape[-1] < self.padded_in_features:
            x = F.pad(x, (0, self.padded_in_features - x.shape[-1]))
        gate_up, _ = self.gate_up_proj(x)
        x = self.act_fn(gate_up)
        x_down, _ = self.down_proj(x)
        if x_down.shape[-1] > self.in_features:
            x_down = x_down[..., : self.in_features]
        return x_down
```

This is only a suggestion, but let me know what you think.

**Comment by johncalesp** [comment]
@Isotr0py if you could review this PR when you get a chance. thank you

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 5]
```diff
@@ -119,6 +143,597 @@
...
+        skip_bias_add: bool = False,
+        params_dtype: torch.dtype | None = None,
+        quant_config: QuantizationConfig | None = None,
+        prefix: str = "",
+        *,
+        return_bias: bool = True,
+        disable_tp: bool = False,
+        loaded_input_size: int | None = None,
+        loaded_output_sizes: list[int] | None = None,
+    ):
```
Comment (Isotr0py) [inline vllm/model_executor/models/qwen2_5_vl.py]: I don't like `WeightPaddingLinear` lintroduction, because it will make the linear implementation too messy.

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 6]
```diff
@@ -282,18 +897,37 @@ def __init__(
...
+                self.in_features,
+                self.hidden_features,
+                self.padded_in_features,
+                self.padded_hidden_features,
+            )
+        self.gate_up_proj = WeightPaddingMergedColumnParallelLinear(
+            input_size=self.padded_in_features,
+            output_sizes=[self.padded_hidden_features] * 2,  # [gate_proj, up_proj]
+            loaded_input_size=in_features,
+            loaded_output_sizes=[hidden_features] * 2,
```
Comment (Isotr0py) [inline vllm/model_executor/models/qwen2_5_vl.py]: I think this will cause regression for unquantized models as well.

**Review by Isotr0py — CHANGES_REQUESTED** [review]
I think this PR is too big (600 lines additions) for a small and Qwen2.5-VL+FP8 specific optimization, especially the `WeightPaddingLinear`.

Actually, you should be able to force CUTLASS kernels through `force_kernel` in `init_fp8_linear_kernel`.
https://github.com/vllm-project/vllm/blob/d247a931cc25e7253feccbd6260d48216ff5c081/vllm/model_executor/kernels/linear/__init__.py#L496-L504

**Comment by huanghua1994** [comment]
Hi @Isotr0py and @johncalesp , much appreciated for your comments and suggestions. I agree that introducing two new classes are too messy and should be avoided. I followed @johncalesp 's suggestion and replaced WeightPadding* classes with `_install_post_pad_hook()`, and only trigger this padding for FP8 scaling. Please review the changes when you have a chance, thank you!

**Comment by huanghua1994** [comment]
@Isotr0py Using `force_kernel` can force `init_fp8_linear_kernel()` to select `CutlassFP8ScaledMMLinearKernel`, but inside `CutlassFP8ScaledMMLinearKernel` it still pads the weights to satisfy the requirement. If we want to avoid padding the weights repeatedly, it is better to pad the weights once in `__init__()`.

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 7]
```diff
@@ -120,6 +124,111 @@
...
+            return
+        padded_module = cast(_Qwen2_5_VLPaddedLinear, module)
+        _pad_weight_per_shard(
+            padded_module,
+            per_shard_output_pad=per_shard_output_pad,
+            input_pad=input_pad,
+        )
+        module._qwen2_5_vl_fp8_weight_padded = True
+
+    qm.process_weights_after_loading = wrapped  # type: ignore[method-assign]
```
Comment (Isotr0py) [inline vllm/model_executor/models/qwen2_5_vl.py:229]: But I think `CutlassFP8ScaledMMLinearKernel` should have already implemented `process_weights_after_loading` to pad weights, which means that we don't have to install another hook for weights padding:
https://github.com/vllm-project/vllm/blob/f2ace1d57d28df8d4c5e973dd62d87f47d628cb3/vllm/model_executor/kernels/linear/scaled_mm/cutlass.py#L173-L212

Comment (huanghua1994) [reply vllm/model_executor/models/qwen2_5_vl.py:229]: Thanks for reminding me this. Yes, you are correct, we don't have to process weights. Then this PR only avoids 2 de-padding and 1 padding to the activation inside the MLP. I redo tests on our target workloads; the performance improvement is small (5% TTFT, the performance results in the PR desc are updated). I defer to your decision regarding whether to close or merge this PR.

**Comment by johncalesp** [comment]
Hi @Isotr0py please let us know if we should keep working on this PR and improve it, thanks!.

**Comment by huanghua1994** [comment]
Please hold the review of this PR. I found some new issues and need to reconsider this PR later.

**Comment by Isotr0py** [comment]
Closing as we can launch CUTLASS kernel for unaligned weights with https://github.com/vllm-project/vllm/pull/42651. Feel free to re-open this if you found there's more optimization can be done :)
